DHT/dhtmem.c

```diff
@@ -16,66 +16,120 @@
 #  include <mem.h>
 #endif /*__BORLANDC__*/
 
+#include "debugging/assert.h"
 #include "dhtvalue.h"
 #include "dhtmem.h"
 #include "dht.h"
 
 typedef unsigned long uLong;
 typedef unsigned char uChar;
 
-static dhtHashValue HashMemoryValue(dhtConstValue v)
+enum {
+  ENSURE_SIZE_OF_ELEMENT_IS_ONE = 1/(1 == sizeof NilMemVal->Data[0])
+};
+
+static dhtHashValue HashMemoryValue(dhtKey k)
 {
-  uLong leng= ((MemVal const *)v)->Leng;
-  uChar const *s= ((MemVal const *)v)->Data;
+  MemVal const * toBeHashed = (MemVal const *)k.value.object_pointer;
+  assert(!!toBeHashed);
+  assert(toBeHashed->Data || !toBeHashed->Leng);
+  uLong leng= toBeHashed->Leng;
+  uChar const *s= toBeHashed->Data;
   dhtHashValue hash= 0;
   uLong i;
-  for (i=0; i<leng; i++) {
-	hash+= s[i];
-	hash+= hash << 10;
-	hash^= hash >> 6;
+  for (i=0; i<leng; ++i) {
+    hash+= s[i];
+    hash+= hash << 10;
+    hash^= hash >> 6;
   }
   hash+= hash << 3;
   hash^= hash >> 11;
   hash+= hash << 15;
   return hash;
 }
-static int EqualMemoryValue(dhtConstValue v1, dhtConstValue v2)
+static int EqualMemoryValue(dhtKey v1, dhtKey v2)
 {
-  if (((MemVal const *)v1)->Leng != ((MemVal const *)v2)->Leng)
-    return 0;
-  if (memcmp(((MemVal const *)v1)->Data, ((MemVal const *)v2)->Data, ((MemVal const *)v1)->Leng))
+  MemVal const * value1 = (MemVal const *)v1.value.object_pointer;
+  MemVal const * value2 = (MemVal const *)v2.value.object_pointer;
+  uLong length;
+  uChar const *data1;
+  uChar const *data2;
+
+  assert(value1 && value2);
+  length = value1->Leng;
+  data1 = value1->Data;
+  assert(data1 || !length);
+  data2 = value2->Data;
+  assert(data2 || !value2->Leng);
+
+  if (length != value2->Leng)
     return 0;
-  else
+  if (!length) // We'll check this here to avoid (maybe) passing NULL as an argument to memcmp.
     return 1;
+  while (length > ((size_t)-1))
+  {
+    if (memcmp(data1, data2, ((size_t)-1)))
```

Comment:
The 3rd argument looks very strange.
Assuming that size_t is an unsigned type, this argument is very big.
Shouldn't we use the (equal) Length value of the two objects?